// @dinamonetworks/hsm-dinamo/dist/hsm/constants/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.M_OF_N_SECRET_LEN = exports.MOD_CORE_M_OF_N_RECOVER = exports.MOD_CORE_M_OF_N_SPLIT = exports.MOD_CORE_HASH_END = exports.SEG_LEN = exports.MAX_HASH_OP_SEG = exports.MOD_CORE_HASH_CONT = exports.MOD_CORE_HASH_START = exports.MOD_CORE_DEL_OBJ = exports.MAX_SYM_OP_SEG = exports.AAP_GCM_MODE_TAG_LEN = exports.AAP_GCM_KEK_IV_LEN = exports.AAP_KEK_IV_LEN = exports.AAP_KEK_IMPORT = exports.AAP_KEK_EXPORT = exports.AAP_XML_FQN_FLAG = exports.AAP_XML_NO_RNS_FLAG = exports.AAP_XML_NOL_FLAG = exports.AAP_FQN_FLAG = exports.AAP_SPKI_ASYM_EXP_FLAG = exports.MAX_RSA_MOD_LEN = exports.MAX_SPKI_LEN = exports.MAX_ECC_SIGN_DER = exports.AAP_MAX_RSA_SIG_LEN = exports.AAP_MAX_RND_SEG = exports.AAP_MAX_JSON = exports.AAP_MAX_HASH_LEN = exports.MAX_OBJ_ID = exports.AAP_MAX_OBJ_ID = exports.AAP_MAX_OBJ_ID_SZ = exports.AAP_DEL_COMMIT = exports.AAP_PAD_CHR = exports.MAX_USR_PWD_LEN = exports.MIN_USR_PWD_LEN = exports.AAP_MAX_USR_PWD = exports.MAX_USR_LEN = exports.MIN_USR_LEN = exports.AAP_MAX_USR_LEN = exports.AAP_USR_PWD_MD_LEN = exports.AAP_CHAP_HASH_LEN = exports.AAP_CHAP_RND_LEN = exports.AAP_HS_ACK_LEN = exports.AAP_HDR_LEN = exports.SIZE_OF_UINT_64 = exports.SIZE_OF_UINT_32 = exports.SIZE_OF_UINT_16 = exports.SIZE_OF_UINT_8 = exports.TEMPORARY_KEY = exports.EXPORTABLE_KEY = exports.HSM_DEFAULT_PORT = void 0;
exports.MOD_CORE_CREATE_KEY = exports.MOD_CORE_EdDSA_VERIFY = exports.MOD_CORE_EdDSA_SIGN = exports.MOD_CORE_ECC_VERIFY2 = exports.MOD_CORE_ECC_SIGN = exports.MOD_CORE_RSA_VERIFY = exports.MOD_CORE_RSA_SIGN = exports.MOD_CORE_HSM_P11_INFO = exports.MOD_CORE_GEN_RAND = exports.MOD_CORE_AUTHENTICATE = exports.MSBO_OP_STATUS = exports.MSBO_OP_UNBLOCK = exports.MSBO_OP_BLOCK = exports.MOD_STAT_BLOCK_OBJ = exports.MSBU_OP_QUERY = exports.MSBU_OP_BLOCK = exports.MSBU_OP_UNBLOCK = exports.MOD_STAT_BLOCK_USR = exports.MOD_STAT_TRUNCATE_LOG = exports.CLOSE_DELAY = exports.NOTIFY_CLOSE = exports.NOTIFY_CID = exports.NOTIFY_ERR = exports.NOTIFY_OK_RM2 = exports.NOTIFY_OK_RM1 = exports.NOTIFY_OK = exports.AAP_SESSION_X509_AUTH = exports.AAP_TLS = exports.SESSION_A_TOKEN_AUTH = exports.AAP_v1 = exports.AAP_HSHAKE_OK_AT = exports.AAP_HSHAKE_OK_NT = exports.AAP_HND_SHAKE = exports.TK_GCM_TAG_LEN = exports.TK_GCM_IV_LEN = exports.SVAULT_MAX_TK_LEN = exports.BUF_IO_LEN = exports.PIX_CSF = exports.XML_SET_CRL = exports.JWS_GEN_OVERHEAD = exports.XML_DSIG_OVERHEAD = exports.P_10_VERSION_1 = exports.P_10_SHA_256 = exports.P_10_MAX_DN = exports.P_10_MAX_DER = exports.MAX_CERT_LEN = exports.MC_M_OF_N_S_MAX = exports.MC_SO_FLAG_FQN = exports.MC_M_OF_N_ES_LEN = exports.MC_M_OF_N_S_PARTLEN = void 0;
exports.MOD_CORE_KEK_WRAP_OBJ_T_P8 = exports.MOD_BCHAIN_PK_EXPORT = exports.MBLKC_PBK_RECOVER = exports.MOD_BCHAIN_VERIFY = exports.MOD_BCHAIN_KINFO = exports.MOD_BCHAIN_EdDSA_VERIFY = exports.MOD_BCHAIN_EdDSA_SIGN = exports.MOD_BCHAIN_HASH = exports.MOD_BCHAIN_BIP32_XPRV_CKD = exports.MOD_BCHAIN_GET_PBK = exports.MOD_BCHAIN_CREATE_PK = exports.MOD_OATH_HOTP_RESYNC = exports.MOD_OATH_GET_OTP = exports.MOD_OATH_CHECK_OTP = exports.MOD_OATH_ISSUE_OBLOB = exports.MOD_SVAULT_SECRET = exports.MOD_SVAULT_MAX_TL = exports.MOD_SVAULT_TOKEN = exports.MOD_PIX_DICT_VERIFY = exports.MOD_PIX_DICT_SIGN = exports.MOD_PIX_JWS_CHECK = exports.MOD_PIX_JWS_GEN = exports.MOD_PIX_VERIFY = exports.MOD_PIX_SIGN = exports.MOD_XML_VERIFY = exports.MOD_XML_SIGN = exports.MOD_CORE_P8_IMPORT = exports.MOD_CORE_P12_EXPORT = exports.MOD_CORE_P8_EXPORT = exports.MOD_CORE_KEK_WRAP_OBJ_IMPORT = exports.MOD_CORE_KEK_WRAP_OBJ_EXPORT = exports.MOD_CORE_FAST_SYM_OP = exports.MOD_CORE_DPK_PROTECT = exports.MOD_CORE_P10_CSR_DN_MAX_LEN = exports.MOD_CORE_P10_CSR_VERSION1 = exports.MOD_CORE_P10_CSR = exports.MOD_CORE_UPDATE_ACL = exports.MOD_CORE_KEK_WRAP_OBJ = exports.MOD_CORE_ASYM_EXP = exports.MOD_CORE_OBJ_EXP = exports.MOD_CORE_GET_OBJ_INFO = exports.MOD_CORE_P8_IMPORT_PWD_LEN = exports.MOD_CORE_P12_IMPORT = exports.MOD_CORE_KEY_IMP = exports.MOD_CORE_A_TOKEN = exports.OP_DELETE_COMMIT = exports.MOD_CORE_CHANGE_PWD = exports.MOD_CORE_DELETE_USR = exports.MOD_CORE_CREATE_USR = exports.MOD_CORE_CREATE_BLOB = void 0;
exports.ERR_INVALID_HASH = exports.ERR_INVALID_SIGNATURE = exports.ERR_CANNOT_TRUNC_LOG = exports.ERR_INVALID_PAYLOAD = exports.ERR_ACCESS_DENIED = exports.ERR_INVALID_OBJ_NAME = exports.ERR_INVALID_SECRET = exports.ERR_NET_FAIL = exports.ERR_NOT_IMPLEMENTED = exports.ERR_CANNOT_OPEN_OBJ = exports.ERR_OBJ_ALREADY_EXISTS = exports.ERR_USR_ALREADY_EXISTS = exports.ERR_USR_NOT_FOUND = exports.ERR_INVALID_KEY = exports.ERR_INVALID_OP = exports.ERR_CANT_ALLOC_R = exports.ERR_INCREMENT = exports.AAP_CHAP_RESP_LEN = exports.MOD_BCHAIN_PK_IMPORT = exports.MOD_BCHAIN_SIGN = exports.MOD_BCHAIN_GET_ADDR = exports.PURE_EdDSA_MODE = exports.MAX_EdDSA_SIG_LEN = exports.MOD_BCHAIN_ABN128_PSC_MULT = exports.MAX_EdDSA_DATA_LEN = exports.MOD_BCHAIN_ABN128_RND_CTX_REFRESH = exports.MOD_BCHAIN_ABN128_RND_CTX_MULT_ADD = exports.MBLKC_ABN128_XY_LEN = exports.ABN128_XY_BUFF_LEN = exports.ABN128_XY_HEX_LEN = exports.MOD_BCHAIN_ABN128_RND_SC = exports.MOD_BCHAIN_ABN128_RND_CTX_GET_PT = exports.BLOCKCHAIN_MAX_DATA_LEN = exports.MBLKC_MAX_HDATA_LEN = exports.MBLKC_EdDSA_MAX_DL = exports.MBLKC_EdDSA_CSTR_UNUSED = exports.MBLKC_MAX_PBK_LEN = exports.MBLKC_MAX_SIG_LEN = exports.BLOCKCHAIN_EdDSA_MAX_CTX_STR_LEN = exports.MBLKC_EdDSA_MAX_CSTRL = exports.MBLKC_EdDSA_SIG_XLM = exports.MBLKC_EdDSA_SIG_SOL = exports.MBLKC_EdDSA_SIG_XRP = exports.MBLKC_EdDSA_SIG_ALGORAND = exports.MBLKC_EdDSA_SIG_Pure = exports.MOD_STAT_SET_NSAUTH_STATE_2 = exports.MOD_STAT_SET_NSAUTH_STATE = exports.MOD_CORE_ASYM_DEC = exports.MOD_CORE_ASYM_ENC = exports.MOD_CORE_ASYM_EXP_FLAG_X509 = void 0;
exports.MOD_CORE_DEL_AUTH_SA_X509 = exports.MOD_CORE_DEL_AUTH_SA_OTP = exports.MOD_CORE_NEW_AUTH_SA = exports.MOD_CORE_SET_TLS_BUNDLE = exports.MOD_CORE_LST_OBJS = exports.MOD_CORE_M_OF_N_RECOVER2 = exports.MOD_CORE_M_OF_N_SPLIT2 = exports.MC_M_OF_N_S_LEN2 = exports.MOD_CORE_M_OF_N_APLIT2 = exports.MC_M_OF_N_S_LEN = exports.MC_M_OF_N_SCHEME_ML = exports.DN_SC_MAX_PIN_LEN = exports.MOD_OATH_OCRA_GEN = exports.MC_AE_FLAG_FQN = exports.MAX_OBJ_V1_RAW_LEN = exports.P8_IMPORT_EXPORT_SECRET_MAX_LEN = exports.P8_IMPORT_EXPORT_SECRET_MIN_LEN = exports.MOD_CORE_P8_EXPORT_PWD_LEN = exports.INTERACTION_VECTOR_LEN = exports.MOD_CORE_ATTR_RAW = exports.LAST_OBJ_LEN_MARKER = exports.TAC_ERROR_BIT = exports.MC_A_TOKEN_INFINITE = exports.MC_A_TOKEN_OP_L_REVOKE = exports.MC_A_TOKEN_OP_L_ISSUE = exports.MC_A_TOKEN_OP_LIST = exports.MC_A_TOKEN_OP_REVOKE = exports.MC_A_TOKEN_OP_ISSUE = exports.A_TOKEN_KLEN = exports.M_OF_N_SECRET_MAX_N = exports.M_OF_N_SECRET_MIN_N = exports.MAX_PASSPHRASE_LEN = exports.MAX_SEED_LEN = exports.BCHAIN_KEY = exports.NOT_COMPRESSED = exports.COMPRESSED = exports.MAX_PK_IMP_LEN = exports.MAX_HASH_LEN = exports.MBLKC_MAX_HRP = exports.MBLKC_CPK_MAX_DATA = exports.SVAULT_MAX_LEN = exports.SVAULT_MIN_LEN = exports.D_SVAULT_T_TKR2 = exports.MAX_OBJ_FQN_LEN = exports.MAX_OBJ_ID_LEN = exports.HSM_ERRORS_MAP = exports.ERR_ACCESS_DENIED_OBJ_BLOCKED = exports.ERR_ACCESS_DENIED_USR_BLOCKED = exports.ERR_CANNOT_CREATE_UNKNOWN_OBJ = exports.ERR_CANNOT_PARSE_DER_PRIV_KEY = void 0;
exports.ATTRIBUTES_OPTIONS = exports.AUP_OPTIONS = exports.OBJ_TYPES = exports.MOD_CORE_GET_ACL = exports.MOD_CORE_LST_BLOBS = exports.MOD_CORE_PROBE_OBJ_INFO = exports.MOD_CORE_DEL_AUTH_SA = void 0;
/**
 * @internal
 */
exports.HSM_DEFAULT_PORT = 4433;
/**
 * @internal
 */
exports.EXPORTABLE_KEY = 0x00000001;
/**
 * @internal
 */
exports.TEMPORARY_KEY = 0x00000004;
/**
 * @internal
 */
exports.SIZE_OF_UINT_8 = 1;
/**
 * @internal
 */
exports.SIZE_OF_UINT_16 = 2;
/**
 * @internal
 */
exports.SIZE_OF_UINT_32 = 4;
/**
 * @internal
 */
exports.SIZE_OF_UINT_64 = 8;
/**
 * @internal
 */
exports.AAP_HDR_LEN = 8;
/**
 * @internal
 */
exports.AAP_HS_ACK_LEN = 4;
/**
 * @internal
 */
exports.AAP_CHAP_RND_LEN = 16;
/**
 * @internal
 */
exports.AAP_CHAP_HASH_LEN = 16;
/**
 * @internal
 */
exports.AAP_USR_PWD_MD_LEN = 16;
/**
 * @internal
 */
exports.AAP_MAX_USR_LEN = 16;
/**
 * Tamanho mínimo do identificador do usuário.
 */
exports.MIN_USR_LEN = 3;
/**
 * Tamanho máximo do identificador do usuário.
 */
exports.MAX_USR_LEN = 16;
/**
 * @internal
 */
exports.AAP_MAX_USR_PWD = 16;
/**
 * Tamanho mínimo da senha do usuário.
 */
exports.MIN_USR_PWD_LEN = 8;
/**
 * Tamanho máximo da senha do usuário.
 */
exports.MAX_USR_PWD_LEN = 16;
/**
 * @internal
 */
exports.AAP_PAD_CHR = 0x00;
/**
 * @internal
 */
exports.AAP_DEL_COMMIT = 1;
/**
 * @internal
 */
exports.AAP_MAX_OBJ_ID_SZ = 32 + 1;
/**
 * @internal
 */
exports.AAP_MAX_OBJ_ID = exports.AAP_MAX_USR_LEN + 1 + exports.AAP_MAX_OBJ_ID_SZ;
/**
 * @internal
 */
exports.MAX_OBJ_ID = 33;
/**
 * @internal
 */
exports.AAP_MAX_HASH_LEN = 64;
/**
 * @internal
 */
exports.AAP_MAX_JSON = 2048;
/**
 * @internal
 */
exports.AAP_MAX_RND_SEG = 2048;
/**
 * @internal
 */
exports.AAP_MAX_RSA_SIG_LEN = 1024;
/**
 * @internal
 */
exports.MAX_ECC_SIGN_DER = 256 + 64;
/**
 * @internal
 */
exports.MAX_SPKI_LEN = 2048;
/**
 * @internal
 */
exports.MAX_RSA_MOD_LEN = 1024;
/**
 * @internal
 */
exports.AAP_SPKI_ASYM_EXP_FLAG = 1 << 31;
/**
 * @internal
 */
exports.AAP_FQN_FLAG = 1 << 31;
/**
 * @internal
 */
exports.AAP_XML_NOL_FLAG = 1 << 30;
/**
 * @internal
 */
exports.AAP_XML_NO_RNS_FLAG = 1 << 29;
/**
 * @internal
 */
exports.AAP_XML_FQN_FLAG = exports.AAP_FQN_FLAG;
/**
 * @internal
 */
exports.AAP_KEK_EXPORT = 1;
/**
 * @internal
 */
exports.AAP_KEK_IMPORT = 2;
/**
 * @internal
 */
exports.AAP_KEK_IV_LEN = 32;
/**
 * @internal
 */
exports.AAP_GCM_KEK_IV_LEN = 16;
/**
 * @internal
 */
exports.AAP_GCM_MODE_TAG_LEN = 16;
/**
 * @internal
 */
exports.MAX_SYM_OP_SEG = 2048;
/**
 * @internal
 */
exports.MOD_CORE_DEL_OBJ = 3;
/**
 * @internal
 */
exports.MOD_CORE_HASH_START = 10;
/**
 * @internal
 */
exports.MOD_CORE_HASH_CONT = 11;
/**
 * @internal
 */
exports.MAX_HASH_OP_SEG = exports.MAX_SYM_OP_SEG;
/**
 * @internal
 */
exports.SEG_LEN = exports.MAX_SYM_OP_SEG;
/**
 * @internal
 */
exports.MOD_CORE_HASH_END = 12;
/**
 * @internal
 */
exports.MOD_CORE_M_OF_N_SPLIT = 78;
/**
 * @internal
 */
exports.MOD_CORE_M_OF_N_RECOVER = 79;
/**
 * Tamanho dos dados do segredo protegidos pelo m de n.
 */
exports.M_OF_N_SECRET_LEN = 24;
/**
 * @internal
 */
exports.MC_M_OF_N_S_PARTLEN = exports.M_OF_N_SECRET_LEN + 3;
/**
 * @internal
 */
exports.MC_M_OF_N_ES_LEN = 32;
/**
 * @internal
 */
exports.MC_SO_FLAG_FQN = 1 << 31;
/**
 * @internal
 */
exports.MC_M_OF_N_S_MAX = 250;
/**
 * @internal
 */
exports.MAX_CERT_LEN = 3072;
/**
 * @internal
 */
exports.P_10_MAX_DER = 10 * 1024;
/**
 * @internal
 */
exports.P_10_MAX_DN = 2048;
/**
 * @internal
 */
exports.P_10_SHA_256 = 3;
/**
 * @internal
 */
exports.P_10_VERSION_1 = 0;
/**
 * @internal
 */
exports.XML_DSIG_OVERHEAD = 7 * 1024;
/**
 * @internal
 */
exports.JWS_GEN_OVERHEAD = 2048;
/**
 * @internal
 */
exports.XML_SET_CRL = 1 << 31;
/**
 * @internal
 */
exports.PIX_CSF = 0x00000000;
/**
 * @internal
 */
exports.BUF_IO_LEN = 512;
/**
 * @internal
 */
exports.SVAULT_MAX_TK_LEN = 256;
/**
 * @internal
 */
exports.TK_GCM_IV_LEN = 16;
/**
 * @internal
 */
exports.TK_GCM_TAG_LEN = 16;
/**
 * @internal
 */
exports.AAP_HND_SHAKE = Number(0x48454c4f);
/**
 * @internal
 */
exports.AAP_HSHAKE_OK_NT = Number(0x4f4b4e54);
/**
 * @internal
 */
exports.AAP_HSHAKE_OK_AT = Number(0x4f4b4154);
/**
 * @internal
 */
exports.AAP_v1 = 1;
/**
 * @internal
 */
exports.SESSION_A_TOKEN_AUTH = 1 << 14;
/**
 * @internal
 */
exports.AAP_TLS = 1;
/**
 * @internal
 */
exports.AAP_SESSION_X509_AUTH = 2;
/**
 * @internal
 */
exports.NOTIFY_OK = 0;
/**
 * @internal
 */
exports.NOTIFY_OK_RM1 = Number(0xfffffffb);
/**
 * @internal
 */
exports.NOTIFY_OK_RM2 = Number(0xfffffff1);
/**
 * @internal
 */
exports.NOTIFY_ERR = Number(0xffffffff);
/**
 * @internal
 */
exports.NOTIFY_CID = 0xfffffff3;
/**
 * @internal
 */
exports.NOTIFY_CLOSE = Number(0xfffffffe);
/**
 * @internal
 */
exports.CLOSE_DELAY = Number(0xffffffff);
/**
 * @internal
 */
exports.MOD_STAT_TRUNCATE_LOG = 505;
/**
 * @internal
 */
exports.MOD_STAT_BLOCK_USR = 507;
/**
 * @internal
 */
exports.MSBU_OP_UNBLOCK = 0;
/**
 * @internal
 */
exports.MSBU_OP_BLOCK = 1;
/**
 * @internal
 */
exports.MSBU_OP_QUERY = 2;
/**
 * @internal
 */
exports.MOD_STAT_BLOCK_OBJ = 514;
/**
 * @internal
 */
exports.MSBO_OP_BLOCK = 0xff;
/**
 * @internal
 */
exports.MSBO_OP_UNBLOCK = 0x00;
/**
 * @internal
 */
exports.MSBO_OP_STATUS = 0xfe;
/**
 * @internal
 */
exports.MOD_CORE_AUTHENTICATE = 17;
/**
 * @internal
 */
exports.MOD_CORE_GEN_RAND = 15;
/**
 * @internal
 */
exports.MOD_CORE_HSM_P11_INFO = 57;
/**
 * @internal
 */
exports.MOD_CORE_RSA_SIGN = 68;
/**
 * @internal
 */
exports.MOD_CORE_RSA_VERIFY = 69;
/**
 * @internal
 */
exports.MOD_CORE_ECC_SIGN = 31;
/**
 * @internal
 */
exports.MOD_CORE_ECC_VERIFY2 = 70;
/**
 * @internal
 */
exports.MOD_CORE_EdDSA_SIGN = 71;
/**
 * @internal
 */
exports.MOD_CORE_EdDSA_VERIFY = 72;
/**
 * @internal
 */
exports.MOD_CORE_CREATE_KEY = 2;
/**
 * @internal
 */
exports.MOD_CORE_CREATE_BLOB = 25;
/**
 * @internal
 */
exports.MOD_CORE_CREATE_USR = 21;
/**
 * @internal
 */
exports.MOD_CORE_DELETE_USR = 22;
/**
 * @internal
 */
exports.MOD_CORE_CHANGE_PWD = 18;
/**
 * @internal
 */
exports.OP_DELETE_COMMIT = 1;
/**
 * @internal
 */
exports.MOD_CORE_A_TOKEN = 61;
/**
 * @internal
 */
exports.MOD_CORE_KEY_IMP = 19;
/**
 * @internal
 */
exports.MOD_CORE_P12_IMPORT = 67;
/**
 * @internal
 */
exports.MOD_CORE_P8_IMPORT_PWD_LEN = 16;
/**
 * @internal
 */
exports.MOD_CORE_GET_OBJ_INFO = 4;
/**
 * @internal
 */
exports.MOD_CORE_OBJ_EXP = 20;
/**
 * @internal
 */
exports.MOD_CORE_ASYM_EXP = 9;
/**
 * @internal
 */
exports.MOD_CORE_KEK_WRAP_OBJ = 42;
/**
 * @internal
 */
exports.MOD_CORE_UPDATE_ACL = 23;
/**
 * @internal
 */
exports.MOD_CORE_P10_CSR = 44;
/**
 * @internal
 */
exports.MOD_CORE_P10_CSR_VERSION1 = 0x00;
/**
 * @internal
 */
exports.MOD_CORE_P10_CSR_DN_MAX_LEN = 2048;
/**
 * @internal
 */
exports.MOD_CORE_DPK_PROTECT = 74;
/**
 * @internal
 */
exports.MOD_CORE_FAST_SYM_OP = 26;
/**
 * @internal
 */
exports.MOD_CORE_KEK_WRAP_OBJ_EXPORT = 1;
/**
 * @internal
 */
exports.MOD_CORE_KEK_WRAP_OBJ_IMPORT = 2;
/**
 * @internal
 */
exports.MOD_CORE_P8_EXPORT = 45;
/**
 * @internal
 */
exports.MOD_CORE_P12_EXPORT = 66;
/**
 * @internal
 */
exports.MOD_CORE_P8_IMPORT = 48;
/**
 * @internal
 */
exports.MOD_XML_SIGN = 5001;
/**
 * @internal
 */
exports.MOD_XML_VERIFY = 5002;
/**
 * @internal
 */
exports.MOD_PIX_SIGN = 90001;
/**
 * @internal
 */
exports.MOD_PIX_VERIFY = 90002;
/**
 * @internal
 */
exports.MOD_PIX_JWS_GEN = 90003;
/**
 * @internal
 */
exports.MOD_PIX_JWS_CHECK = 90004;
/**
 * @internal
 */
exports.MOD_PIX_DICT_SIGN = 90005;
/**
 * @internal
 */
exports.MOD_PIX_DICT_VERIFY = 90006;
/**
 * @internal
 */
exports.MOD_SVAULT_TOKEN = 100001;
/**
 * @internal
 */
exports.MOD_SVAULT_MAX_TL = 256;
/**
 * @internal
 */
exports.MOD_SVAULT_SECRET = 100002;
/**
 * @internal
 */
exports.MOD_OATH_ISSUE_OBLOB = 70001;
/**
 * @internal
 */
exports.MOD_OATH_CHECK_OTP = 70002;
/**
 * @internal
 */
exports.MOD_OATH_GET_OTP = 70007;
/**
 * @internal
 */
exports.MOD_OATH_HOTP_RESYNC = 70005;
/**
 * @internal
 */
exports.MOD_BCHAIN_CREATE_PK = 120505;
/**
 * @internal
 */
exports.MOD_BCHAIN_GET_PBK = 120501;
/**
 * @internal
 */
exports.MOD_BCHAIN_BIP32_XPRV_CKD = 120507;
/**
 * @internal
 */
exports.MOD_BCHAIN_HASH = 120510;
/**
 * @internal
 */
exports.MOD_BCHAIN_EdDSA_SIGN = 120511;
/**
 * @internal
 */
exports.MOD_BCHAIN_EdDSA_VERIFY = 120512;
/**
 * @internal
 */
exports.MOD_BCHAIN_KINFO = 120506;
/**
 * @internal
 */
exports.MOD_BCHAIN_VERIFY = 120509;
/**
 * @internal
 */
exports.MBLKC_PBK_RECOVER = 0xff;
/**
 * @internal
 */
exports.MOD_BCHAIN_PK_EXPORT = 120504;
/**
 * @internal
 */
exports.MOD_CORE_KEK_WRAP_OBJ_T_P8 = 0xffffffdf;
/**
 * @internal
 */
exports.MOD_CORE_ASYM_EXP_FLAG_X509 = 1 << 31;
/**
 * @internal
 */
exports.MOD_CORE_ASYM_ENC = 7;
/**
 * @internal
 */
exports.MOD_CORE_ASYM_DEC = 8;
/**
 * @internal
 */
exports.MOD_STAT_SET_NSAUTH_STATE = 533;
/**
 * @internal
 */
exports.MOD_STAT_SET_NSAUTH_STATE_2 = 536;
/**
 * @internal
 */
exports.MBLKC_EdDSA_SIG_Pure = 254;
/**
 * @internal
 */
exports.MBLKC_EdDSA_SIG_ALGORAND = 253;
/**
 * @internal
 */
exports.MBLKC_EdDSA_SIG_XRP = 252;
/**
 * @internal
 */
exports.MBLKC_EdDSA_SIG_SOL = 251;
/**
 * @internal
 */
exports.MBLKC_EdDSA_SIG_XLM = 250;
/**
 * @internal
 */
exports.MBLKC_EdDSA_MAX_CSTRL = 255 + 1;
/**
 * Tamanho máximo da string de contexto para assinatura EdDSA.
 */
exports.BLOCKCHAIN_EdDSA_MAX_CTX_STR_LEN = 255; // Some value of MBLKC_EdDSA_MAX_CSTRL
/**
 * @internal
 */
exports.MBLKC_MAX_SIG_LEN = 128;
/**
 * @internal
 */
exports.MBLKC_MAX_PBK_LEN = 128;
/**
 * @internal
 */
exports.MBLKC_EdDSA_CSTR_UNUSED = "";
/**
 * @internal
 */
exports.MBLKC_EdDSA_MAX_DL = 4 * 1024 * 1024;
/**
 * @internal
 */
exports.MBLKC_MAX_HDATA_LEN = 2 * 1024 * 1024;
/**
 * Tamanho máximo do bloco de dados.
 */
exports.BLOCKCHAIN_MAX_DATA_LEN = 2097152; // The same value of MBLKC_MAX_HDATA_LEN
/**
 * @internal
 */
exports.MOD_BCHAIN_ABN128_RND_CTX_GET_PT = 120519;
/**
 * @internal
 */
exports.MOD_BCHAIN_ABN128_RND_SC = 120514;
/**
 * Tamanho máximo dos pontos X e Y em hexadecimal.
 */
exports.ABN128_XY_HEX_LEN = 64;
/**
 * Tamanho máximo dos pontos X e Y em buffer.
 */
exports.ABN128_XY_BUFF_LEN = 32;
/**
 * @internal
 */
exports.MBLKC_ABN128_XY_LEN = exports.ABN128_XY_BUFF_LEN;
/**
 * @internal
 */
exports.MOD_BCHAIN_ABN128_RND_CTX_MULT_ADD = 120520;
/**
 * @internal
 */
exports.MOD_BCHAIN_ABN128_RND_CTX_REFRESH = 120518;
/**
 * @internal
 */
exports.MAX_EdDSA_DATA_LEN = 4 * 1024 * 1024;
/**
 * @internal
 */
exports.MOD_BCHAIN_ABN128_PSC_MULT = 120515;
/**
 * @internal
 */
exports.MAX_EdDSA_SIG_LEN = 256;
/**
 * @internal
 */
exports.PURE_EdDSA_MODE = 0;
/**
 * @internal
 */
exports.MOD_BCHAIN_GET_ADDR = 120502;
/**
 * @internal
 */
exports.MOD_BCHAIN_SIGN = 120508;
/**
 * @internal
 */
exports.MOD_BCHAIN_PK_IMPORT = 120503;
/**
 * @internal
 */
exports.AAP_CHAP_RESP_LEN = exports.AAP_CHAP_RND_LEN + exports.AAP_MAX_USR_LEN;
/**
 * @internal
 */
exports.ERR_INCREMENT = 0x7fffffff;
/**
 * @internal
 */
exports.ERR_CANT_ALLOC_R = 0x8000138e & exports.ERR_INCREMENT;
/**
 * @internal
 */
exports.ERR_INVALID_OP = 0x80001390 & exports.ERR_INCREMENT;
/**
 * @internal
 */
exports.ERR_INVALID_KEY = 0x80001391 & exports.ERR_INCREMENT;
/**
 * @internal
 */
exports.ERR_USR_NOT_FOUND = 0x8000139c & exports.ERR_INCREMENT;
/**
 * @internal
 */
exports.ERR_USR_ALREADY_EXISTS = 0x80001395 & exports.ERR_INCREMENT;
/**
 * @internal
 */
exports.ERR_OBJ_ALREADY_EXISTS = 0x8000139e & exports.ERR_INCREMENT;
/**
 * @internal
 */
exports.ERR_CANNOT_OPEN_OBJ = 0x8000138c & exports.ERR_INCREMENT;
/**
 * @internal
 */
exports.ERR_NOT_IMPLEMENTED = 0x8000139b & exports.ERR_INCREMENT;
/**
 * @internal
 */
exports.ERR_NET_FAIL = 0x80001389 & exports.ERR_INCREMENT;
/**
 * @internal
 */
exports.ERR_INVALID_SECRET = 0x800013f5 & exports.ERR_INCREMENT;
/**
 * @internal
 */
exports.ERR_INVALID_OBJ_NAME = 0x8000139f & exports.ERR_INCREMENT;
/**
 * @internal
 */
exports.ERR_ACCESS_DENIED = 0x8000138a & exports.ERR_INCREMENT;
/**
 * @internal
 */
exports.ERR_INVALID_PAYLOAD = 0x8000139d & exports.ERR_INCREMENT;
/**
 * @internal
 */
exports.ERR_CANNOT_TRUNC_LOG = 0x800013ab & exports.ERR_INCREMENT;
/**
 * @internal
 */
exports.ERR_INVALID_SIGNATURE = 0x800013f3 & exports.ERR_INCREMENT;
/**
 * @internal
 */
exports.ERR_INVALID_HASH = 0x800015a1 & exports.ERR_INCREMENT;
/**
 * @internal
 */
exports.ERR_CANNOT_PARSE_DER_PRIV_KEY = 0x800013c3 & exports.ERR_INCREMENT;
/**
 * @internal
 */
exports.ERR_CANNOT_CREATE_UNKNOWN_OBJ = 0x80001583 & exports.ERR_INCREMENT;
/**
 * @internal
 */
exports.ERR_ACCESS_DENIED_USR_BLOCKED = 0x80001400 & exports.ERR_INCREMENT;
/**
 * @internal
 */
exports.ERR_ACCESS_DENIED_OBJ_BLOCKED = 0x80001404 & exports.ERR_INCREMENT;
/**
 * @internal
 */
exports.HSM_ERRORS_MAP = new Map();
exports.HSM_ERRORS_MAP.set(exports.ERR_CANT_ALLOC_R, "ERR_CANT_ALLOC_R");
exports.HSM_ERRORS_MAP.set(exports.ERR_INVALID_OP, "ERR_INVALID_OP");
exports.HSM_ERRORS_MAP.set(exports.ERR_INVALID_KEY, "ERR_INVALID_KEY");
exports.HSM_ERRORS_MAP.set(exports.ERR_USR_NOT_FOUND, "ERR_USR_NOT_FOUND");
exports.HSM_ERRORS_MAP.set(exports.ERR_USR_ALREADY_EXISTS, "ERR_USR_ALREADY_EXISTS");
exports.HSM_ERRORS_MAP.set(exports.ERR_OBJ_ALREADY_EXISTS, "ERR_OBJ_ALREADY_EXISTS");
exports.HSM_ERRORS_MAP.set(exports.ERR_CANNOT_OPEN_OBJ, "ERR_CANNOT_OPEN_OBJ");
exports.HSM_ERRORS_MAP.set(exports.ERR_NOT_IMPLEMENTED, "ERR_NOT_IMPLEMENTED");
exports.HSM_ERRORS_MAP.set(exports.ERR_NET_FAIL, "ERR_NET_FAIL");
exports.HSM_ERRORS_MAP.set(exports.ERR_INVALID_SECRET, "ERR_INVALID_SECRET");
exports.HSM_ERRORS_MAP.set(exports.ERR_INVALID_OBJ_NAME, "ERR_INVALID_OBJ_NAME");
exports.HSM_ERRORS_MAP.set(exports.ERR_ACCESS_DENIED, "ERR_ACCESS_DENIED");
exports.HSM_ERRORS_MAP.set(exports.ERR_INVALID_PAYLOAD, "ERR_INVALID_PAYLOAD");
exports.HSM_ERRORS_MAP.set(exports.ERR_CANNOT_TRUNC_LOG, "ERR_CANNOT_TRUNC_LOG");
exports.HSM_ERRORS_MAP.set(exports.ERR_INVALID_SIGNATURE, "ERR_INVALID_SIGNATURE");
exports.HSM_ERRORS_MAP.set(exports.ERR_INVALID_HASH, "ERR_INVALID_HASH");
/**
 * @internal
 */
exports.MAX_OBJ_ID_LEN = exports.MAX_USR_LEN + 1 + exports.MAX_OBJ_ID;
/**
 * @internal
 */
exports.MAX_OBJ_FQN_LEN = exports.MAX_OBJ_ID_LEN;
/**
 * @internal
 */
exports.D_SVAULT_T_TKR2 = 1;
/**
 * @internal
 * Tamanho mínimo da secret para tokenização.
 * @type {number}
 */
exports.SVAULT_MIN_LEN = 6;
/**
 * @internal
 * Tamanho máximo da secret para tokenização.
 * @type {number}
 */
exports.SVAULT_MAX_LEN = 256;
/**
 * @internal
 */
exports.MBLKC_CPK_MAX_DATA = 1152;
/**
 * @internal
 */
exports.MBLKC_MAX_HRP = 64;
/**
 * @internal
 */
exports.MAX_HASH_LEN = 64;
/**
 * @internal
 */
exports.MAX_PK_IMP_LEN = 1024;
/**
 * @internal
 */
exports.COMPRESSED = 1;
/**
 * @internal
 */
exports.NOT_COMPRESSED = 0;
/**
 * @internal
 */
exports.BCHAIN_KEY = 0x00004000;
/**
 * @internal
 */
exports.MAX_SEED_LEN = 765;
/**
 * @internal
 */
exports.MAX_PASSPHRASE_LEN = 255;
/**
 * Tamanho mínimo de n.
 */
exports.M_OF_N_SECRET_MIN_N = 2;
/**
 * Tamanho máximo de n.
 */
exports.M_OF_N_SECRET_MAX_N = 250;
/**
 * @internal
 */
exports.A_TOKEN_KLEN = 16;
/**
 * @internal
 */
exports.MC_A_TOKEN_OP_ISSUE = 1;
/**
 * @internal
 */
exports.MC_A_TOKEN_OP_REVOKE = 2;
/**
 * @internal
 */
exports.MC_A_TOKEN_OP_LIST = 3;
/**
 * @internal
 */
exports.MC_A_TOKEN_OP_L_ISSUE = 4;
/**
 * @internal
 */
exports.MC_A_TOKEN_OP_L_REVOKE = 5;
/**
 * @internal
 */
exports.MC_A_TOKEN_INFINITE = ~0;
/**
 * @internal
 */
exports.TAC_ERROR_BIT = 0x1 << 31;
/**
 * @internal
 */
exports.LAST_OBJ_LEN_MARKER = exports.MC_A_TOKEN_INFINITE & ~exports.TAC_ERROR_BIT;
/**
 * @internal
 */
exports.MOD_CORE_ATTR_RAW = 1 << 31;
/**
 * @internal
 */
exports.INTERACTION_VECTOR_LEN = 32;
/**
 * @internal
 */
exports.MOD_CORE_P8_EXPORT_PWD_LEN = 16;
/**
 * Tamanho mínimo da senha para exportação de chave PKCS#8.
 */
exports.P8_IMPORT_EXPORT_SECRET_MIN_LEN = 8;
/**
 * Tamanho máximo da senha para exportação de chave PKCS#8.
 */
exports.P8_IMPORT_EXPORT_SECRET_MAX_LEN = 16;
/**
 * Tamanho máximo do material da chave para importação.
 */
exports.MAX_OBJ_V1_RAW_LEN = 4096;
/**
 * @internal
 */
exports.MC_AE_FLAG_FQN = 1 << 31;
/**
 * @internal
 */
exports.MOD_OATH_OCRA_GEN = 70008;
/**
 * @internal
 */
exports.DN_SC_MAX_PIN_LEN = 8;
/**
 * @internal
 */
exports.MC_M_OF_N_SCHEME_ML = 3;
/**
 * @internal
 */
exports.MC_M_OF_N_S_LEN = 24;
/**
 * @internal
 */
exports.MOD_CORE_M_OF_N_APLIT2 = 89;
/**
 * @internal
 */
exports.MC_M_OF_N_S_LEN2 = 56;
/**
 * @internal
 */
exports.MOD_CORE_M_OF_N_SPLIT2 = 89;
/**
 * @internal
 */
exports.MOD_CORE_M_OF_N_RECOVER2 = 90;
/**
 * @internal
 */
exports.MOD_CORE_LST_OBJS = 16;
/**
 * @internal
 */
exports.MOD_CORE_SET_TLS_BUNDLE = 63;
/**
 * @internal
 */
exports.MOD_CORE_NEW_AUTH_SA = 62;
/**
 * @internal
 */
exports.MOD_CORE_DEL_AUTH_SA_OTP = 17;
/**
 * @internal
 */
exports.MOD_CORE_DEL_AUTH_SA_X509 = 18;
/**
 * @internal
 */
exports.MOD_CORE_DEL_AUTH_SA = 51;
/**
 * @internal
 */
exports.MOD_CORE_PROBE_OBJ_INFO = 73;
/**
 * @internal
 */
exports.MOD_CORE_LST_BLOBS = 87;
/**
 * @internal
 */
exports.MOD_CORE_GET_ACL = 30;
/**
 * @internal
 */
exports.OBJ_TYPES = {
    "INVALID_OBJ_TYPE": 0,
    "DES": 1,
    "3DES_112": 2,
    "3DES_168": 3,
    "RSA_512": 4,
    "RSA_1024": 5,
    "RSA_2048": 6,
    "AES_128": 7,
    "AES_192": 8,
    "AES_256": 9,
    "ARC4": 10,
    "RSA_4096": 11,
    "OBJ_BLOB": 12,
    "OBJ_BLOB_X509": 13,
    "OBJ_BLOB_PKCS7": 14,
    "OBJ_BLOB_CRL": 15,
    "OBJ_BLOB_HOTP": 16,
    "ECC_SECP112R1": 18,
    "ECC_SECP112R2": 19,
    "ECC_X9_62_PRIME192V1": 19,
    "ECC_SECP128R1": 20,
    "ECC_SECP128R2": 21,
    "ECC_SECP160K1": 22,
    "ECC_SECP160R1": 23,
    "ECC_SECP160R2": 24,
    "ECC_SECP192K1": 25,
    "ECC_SECP192R1": 26,
    "ECC_SECP224K1": 27,
    "ECC_SECP224R1": 28,
    "ECC_SECP256K1": 29,
    "ECC_SECP256R1": 30,
    "ECC_X9_62_PRIME256V1": 30,
    "ECC_SECP384R1": 31,
    "ECC_SECP521R1": 32,
    "ECC_X9_62_PRIME192V2": 33,
    "ECC_X9_62_PRIME192V3": 34,
    "ECC_X9_62_PRIME239V1": 35,
    "ECC_X9_62_PRIME239V2": 36,
    "ECC_X9_62_PRIME239V3": 37,
    "ECC_BRAINPOOL_P160R1": 38,
    "ECC_BRAINPOOL_P160T1": 39,
    "ECC_BRAINPOOL_P192R1": 40,
    "ECC_BRAINPOOL_P192T1": 41,
    "ECC_BRAINPOOL_P224R1": 42,
    "ECC_BRAINPOOL_P224T1": 43,
    "ECC_BRAINPOOL_P256R1": 44,
    "ECC_BRAINPOOL_P256T1": 45,
    "ECC_BRAINPOOL_P320R1": 46,
    "ECC_BRAINPOOL_P320T1": 47,
    "ECC_BRAINPOOL_P384R1": 48,
    "ECC_BRAINPOOL_P384T1": 49,
    "ECC_BRAINPOOL_P512R1": 50,
    "ECC_BRAINPOOL_P512T1": 51,
    "ARC5": 89,
    "MAP_2_OBJ": 90,
    "DESX": 91,
    'HMAC_MD5': 92,
    'HMAC_SHA1': 93,
    'HMAC_SHA2_256': 94,
    'HMAC_SHA2_384': 95,
    'HMAC_SHA2_512': 96,
    'HMAC_SHA2_224': 97,
    'HMAC_SHA3_224': 98,
    'HMAC_SHA3_256': 99,
    'HMAC_SHA3_384': 100,
    'HMAC_SHA3_512': 101,
    'CMAC_AES': 250,
    'CMAC_DES': 251,
    "PKCS11_MAP": 120,
    "RSA_1152": 121,
    "RSA_1408": 122,
    "RSA_1984": 123,
    "RSA_8192": 124,
    "EXT_MAP_2_OBJ": 125,
    "RSA_2304": 126,
    "RSA_2560": 127,
    "RSA_2816": 128,
    "RSA_3072": 129,
    "RSA_1976": 130,
    "RSA_1536": 131,
    "DSA_2048": 150,
    "DSA_3072": 151,
    "DH_RFC5114_2048": 152,
    "DH_RFC7919_2048": 153,
    "DH_RFC7919_3072": 154,
    'OBJ_CSP': 193,
    "ECX_ED25519": 201,
    "ECX_ED448": 202,
    "ECX_X25519": 203,
    "ECX_X448": 204,
    'OBJ_PUBKEY_BLOB': 350,
    "OBJ_PUBKEY_RSA_BLOB": 351,
    "OBJ_PUBKEY_ECC_BLOB": 352,
    "OBJ_PUBKEY_SPKI_RSA_BLOB": 353,
    "OBJ_PUBKEY_ECC_Ed_BLOB": 354,
    "OBJ_PUBKEY_ECC_X_BLOB": 355,
    "OBJ_PUBKEY_DSA_BLOB": 356,
    "OBJ_PUBKEY_DH_BLOB": 357,
    "BLKC_BTC_MINI_KEY": 1771,
    "BLKC_BIP32_SXPRIV": 1772,
    "BLKC_BIP39_SXPRIV": 1773,
    "BLKC_BIP32_XPRIV": 1774,
    "BLKC_BIP32_Ed25519_XPRIV": 1775,
    "ALT_BN128": 2771,
    "VIRTUAL_SPB_CERT_BLOB": 9437,
    "VIRTUAL_SPB_CERT_BLOB2": 9438,
    "OBJ_DS_BIND_BLOB": 9439
};
/**
 * @internal
 */
exports.AUP_OPTIONS = {
    DIG_SIG: 1,
    DATA_CRYPTO: 2,
    KeK: 3,
    MAC: 4,
    KDF: 5,
    CRYPTOGRAM: 6,
    KEY_TRANSLATE: 7,
    EFT_CVK: 8,
    EFT_VISA_PVK: 9,
    EFT_IBM_3624: 10,
    EFT_PEK: 11,
    EFT_BDK: 12,
    EFT_IPEK: 13,
    EMV_IMK: 14,
    EMV_IMKDAC: 15,
    EMV_IMKENC: 16,
    EMV_IMKMAC: 17,
    EMV_KeK: 18,
    EMV_IMKKDF: 19,
    EMV_IMKACRYPTO: 20,
    EFT_KeK: 21,
    EMV_DIG_SIG: 22,
    EFT_TR31_KBPK: 23,
    EFT_TR34_PK: 24,
    SPB_PK: 25,
    AUP_MASK: 0xFF000000,
    AUP_ANY: 0,
    AUP_RESERVED: 0xFF,
};
/**
 * @internal
 */
exports.ATTRIBUTES_OPTIONS = {
    OBJ_NOP: 0x00,
    BCHAIN_KEY: 0x4000,
    EXPORTABLE_KEY: 0x01,
    TEMPORARY_KEY: 0x04,
    OATH: 0x10,
    OATH_USE_TIME: 0x80,
    PKCS11: 0x400,
    RSA_PUB_EXP3: 0x800,
    SPB: 0x1000,
    HSM_GEN: 0x8000,
    TSP: 0x2000
};
