// @dinamonetworks/hsm-dinamo/dist/hsm/exceptions/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RequiredParameterError = exports.InvalidKeyNameError = exports.InvalidParameterError = exports.HsmError = exports.AuthModeNotSpecifiedError = exports.HsmOptionError = exports.ConnectionError = void 0;
const connection_error_1 = __importDefault(require("./connection_error"));
exports.ConnectionError = connection_error_1.default;
const hsm_option_error_1 = __importDefault(require("./hsm_option_error"));
exports.HsmOptionError = hsm_option_error_1.default;
const auth_mode_not_spacified_error_1 = __importDefault(require("./auth_mode_not_spacified_error"));
exports.AuthModeNotSpecifiedError = auth_mode_not_spacified_error_1.default;
const hsm_error_1 = __importDefault(require("./hsm_error"));
exports.HsmError = hsm_error_1.default;
const invalid_parameter_error_1 = __importDefault(require("./invalid_parameter_error"));
exports.InvalidParameterError = invalid_parameter_error_1.default;
const invalid_key_name_error_1 = __importDefault(require("./invalid_key_name_error"));
exports.InvalidKeyNameError = invalid_key_name_error_1.default;
const required_parameter_error_1 = __importDefault(require("./required_parameter_error"));
exports.RequiredParameterError = required_parameter_error_1.default;
